3_train/src/models.py

```diff
@@ -0,0 +1,351 @@
+        If True, expects the batch inputs to be of shape [batch, seq, features] otherwise, the
+        shape has to be [seq, batch, features], by default True.
+    initial_forget_bias : int, optional
+        Value of the initial forget gate bias, by default 0
+    """
+
+    def __init__(self,
+                 input_size: int,
+                 hidden_size: int,
+                 batch_first: bool = True,
+                 initial_forget_bias: int = 0):
+        super(LSTM, self).__init__()
+
+        self.input_size = input_size
+        self.hidden_size = hidden_size
+        self.batch_first = batch_first
+        self.initial_forget_bias = initial_forget_bias
+
+        # create tensors of learnable parameters
+        self.weight_ih = nn.Parameter(torch.FloatTensor(input_size, 4 * hidden_size))
+        self.weight_hh = nn.Parameter(torch.FloatTensor(hidden_size, 4 * hidden_size))
+        self.bias = nn.Parameter(torch.FloatTensor(4 * hidden_size))
+
+        # initialize parameters
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        """Initialize all learnable parameters of the LSTM"""
+        nn.init.orthogonal_(self.weight_ih.data)
+
+        weight_hh_data = torch.eye(self.hidden_size)
+        weight_hh_data = weight_hh_data.repeat(1, 4)
+        self.weight_hh.data = weight_hh_data
+        nn.init.constant_(self.bias.data, val=0)
+
+        if self.initial_forget_bias != 0:
+            self.bias.data[:self.hidden_size] = self.initial_forget_bias
+
+    def forward(self, x: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """[summary]
+        
+        Parameters
+        ----------
+        x : torch.Tensor
+            Tensor, containing a batch of input sequences. Format must match the specified format,
+            defined by the batch_first agrument.
+
+        Returns
+        -------
+        h_n : torch.Tensor
+            The hidden states of each time step of each sample in the batch.
+        c_n : torch.Tensor]
+            The cell states of each time step of each sample in the batch.
+        """
+        if self.batch_first:
+            x = x.transpose(0, 1)
+
+        seq_len, batch_size, _ = x.size()
+
+        h_0 = x.data.new(batch_size, self.hidden_size).zero_()
+        c_0 = x.data.new(batch_size, self.hidden_size).zero_()
+        h_x = (h_0, c_0)
+
+        # empty lists to temporally store all intermediate hidden/cell states
+        h_n, c_n = [], []
+
+        # expand bias vectors to batch size
+        bias_batch = (self.bias.unsqueeze(0).expand(batch_size, *self.bias.size()))
+
+        # perform forward steps over input sequence
+        for t in range(seq_len):
+            h_0, c_0 = h_x
+
+            # calculate gates
+            gates = (torch.addmm(bias_batch, h_0, self.weight_hh) + torch.mm(x[t], self.weight_ih))
+            f, i, o, g = gates.chunk(4, 1)
+
+            c_1 = torch.sigmoid(f) * c_0 + torch.sigmoid(i) * torch.tanh(g)
```

Comment:
the "_1" notation is a little strange, I was tempted to suggest that it be "_t" but I guess it makes since if you think of it relative to the last data point since you're always starting with "_0" as well (i.e., it's not truly the 0th, it's the last one)

Reply:
Agreed on all counts! Again, I'll leave the author's notation as is. In the past I've used `x_i` for the current, `x_p` for the next (i plus one), and `x_m` for the previous (i minus one), but wasn't totally happy with that system because I don't think it's immediately apparent to a new reader what it means, and because the p in `x_p` could be misinterpreted as "previous".